### PR TITLE
add support for using request header values

### PR DESCRIPTION
**Pull request by david-peden-q2**
This PR allows rules to have access to request headers.

### CODE REVIEW

**Comment by muratg** [comment]
@jkotalik / @Tratcher could you guys review?

File: src/Microsoft.AspNetCore.Rewrite/Internal/IISUrlRewrite/InputParser.cs [hunk 1]
```diff
@@ -136,6 +136,17 @@ private void ParseParameter(ParserContext context, IList<PatternSegment> results
                                 }
                                 return;
                             }
+                        case "RequestHeader":
+                        {
+                            Pattern pattern = ParseString(context, uriMatchPart);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Rewrite/Internal/IISUrlRewrite/InputParser.cs]: nit: var

File: test/Microsoft.AspNetCore.Rewrite.Tests/IISUrlRewrite/InputParserTests.cs [hunk 2]
```diff
@@ -75,6 +75,18 @@ public void EvaluatUriEncodeRule(string testString, string expected)
         }
 
         [Theory]
+        [InlineData("hey/{RequestHeader:Tenant-Id}", "hey/805ffafa-293d-427b-a136-96f196087667")]
+        public void EvaluateRequestHeaderRule(string testString, string expected)
+        {
+            const string tenantId = "805ffafa-293d-427b-a136-96f196087667";
+            var pattern = new InputParser().ParseInputString(testString, UriMatchPart.Path);
+            RewriteContext rewriteContext = CreateTestRewriteContext();
```
Comment (jkotalik) [inline test/Microsoft.AspNetCore.Rewrite.Tests/IISUrlRewrite/InputParserTests.cs]: nit: var

File: test/Microsoft.AspNetCore.Rewrite.Tests/IISUrlRewrite/InputParserTests.cs [hunk 3]
```diff
@@ -75,6 +75,18 @@ public void EvaluatUriEncodeRule(string testString, string expected)
         }
 
         [Theory]
+        [InlineData("hey/{RequestHeader:Tenant-Id}", "hey/805ffafa-293d-427b-a136-96f196087667")]
+        public void EvaluateRequestHeaderRule(string testString, string expected)
```
Comment (jkotalik) [inline test/Microsoft.AspNetCore.Rewrite.Tests/IISUrlRewrite/InputParserTests.cs]: Add a few negative test cases here.

File: src/Microsoft.AspNetCore.Rewrite/Internal/PatternSegments/HeaderSegment.cs [hunk 4]
```diff
@@ -1,20 +1,44 @@
...
+            string header;
+            if (_pattern != null)
+            {
+                // PERF 
+                // Because we need to be able to evaluate multiple nested patterns,
+                // we provided a new string builder and evaluate the new pattern,
+                // and restore it after evaluation.
+                var oldBuilder = context.Builder;
+                context.Builder = new StringBuilder(64);
+                header = _pattern.Evaluate(context, ruleBackReferences, conditionBackReference);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/PatternSegments/HeaderSegment.cs]: What is the pattern scenario? You didn't add a test for it did you?

**Comment by davidpeden3** [comment]
@jkotalik @Tratcher i should have some time to look at this tomorrow. thanks

**Comment by muratg** [comment]
Thanks @davidpeden3 for the PR. Closing this as there's no recent activity.

Please file a bug if you'd like to bring this back in.

**Pull request closed without merging**